Fixes #10462 - task for creating hosts

tstrachota opened the pull request.

Commit:
Fixes #10462 - task for creating hosts

tstrachota commented:
@ShimShtein rebased onto current master

ShimShtein commented on app/lib/foreman_deployments/tasks/host_creation_task_definition.rb:
```diff
@@ -0,0 +1,39 @@
+        if object.is_a? Host::Managed
+          object_params['build'] ||= true
+          # Set attributes from a hostgroup
+          object.attributes = object.apply_inherited_attributes(object_params)
```
You don't have to call `apply_inherited_attributes` explicitly, if the host is a new object. It is called implicitly [here](https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L48).
I can thank @domcleal for this change - he insisted on a seamless integration of this feature.

ShimShtein commented on app/lib/foreman_deployments/tasks/wait_until_built_task_definition.rb:
```diff
@@ -0,0 +1,60 @@
+        output_hash
+      end
+
+      def self.build_status(host = nil)
```
Can you please add some unit testing to this method?

tstrachota replied:
done

ShimShtein commented on app/lib/foreman_deployments/tasks/wait_until_built_task_definition.rb:
```diff
@@ -0,0 +1,60 @@
+      end
+
+      def preliminary_output(_parameters)
+        WaitUntilBuiltTaskDefinition.create_output(Host::Managed.new)
```
It's a bit dangerous to output a new instance of a host as an output for the task. The record is not valid, so if someone will try to validate this model down the line, the validation will fail resulting task validation failure.

tstrachota replied:
I couldn't find any better alternative of the preliminary output for this task. Any ideas?

ShimShtein replied:
Probably a bad one: try to find a host which is already built. In a mature system it's not such a bad idea, but for empty systems it will become a problem. Additionally it adds kind of non-deterministic behavior to the system.
Let's leave this as is for now, and think about it later.

ShimShtein commented on app/lib/foreman_deployments/tasks/wait_until_built_task_definition.rb:
```diff
@@ -0,0 +1,60 @@
+        end
+
+        def poll_external_task
+          # TODO: check the case when no host_id is sent
```
We can put this input validation into the `#validate` method. This task cannot run without it.

tstrachota replied:
Done. I added the same check to the dynflow action itself too, to catch the cases when nil would be propagated due to a failure in a previous task.

ShimShtein commented on test/unit/lib/tasks/creation_task_definition_test.rb:
```diff
@@ -0,0 +1,57 @@
+      assert_equal(ForemanDeployments::Tasks::HostCreationTaskDefinition, definition.class)
+    end
+
+    test 'it returns instance of HostCreationTaskDefinition for other classes' do
```
Copy - paste typo: "it returns instance of **HostCreationTaskDefinition** for..." should be "CreationTaskDefinition"

Commit:
Refs #10462 - more unit tests and validations

tstrachota commented:
@ShimShtein updated & rebased

Commit:
Refs #10462 - search task fixes

Parameter 'klass' renamed to 'class'. There was actually no reason for using
the modified name since the parameter is string.

Another fix is in the tag name.

ShimShtein commented on test/unit/lib/tasks/wait_until_built_task_definition_test.rb:
```diff
@@ -0,0 +1,71 @@
+      assert_equal(false, status['build'])
+    end
+
+    test 'is negative for host with two reports, where the last succeeded' do
```
Typo: "is **positive** ~~negative~~ for host..."

ShimShtein commented on app/lib/foreman_deployments/tasks/wait_until_built_task_definition.rb:
```diff
@@ -0,0 +1,65 @@
+      end
+
+      def self.build_status(host = nil)
+        status = (!host.nil? && (host.reports.count > 1) && !host.reports.last.error?)
```
Do we need to adapt this code to the new status design in core?

tstrachota replied:
Nope, status changes don't affect it at all. There might be some slight changes after the reports STI is merged.

ShimShtein commented on app/lib/foreman_deployments/tasks/creation_task_definition.rb:
```diff
@@ -18,27 +18,31 @@ def validate
 
       def preliminary_output
-        CreationTaskDefinition.create_output(CreationTaskDefinition.create_object(parameters.configured))
+        self.class.create_output(self.class.create_object(parameters.configured))
```
The same needs to be done for `#validate`

Commit:
Refs #10462 - validation fixes

- fixed interfaces initialization in host create task
- calling build_object from correct class in creation task's validate
- fixed passing params in Value#build

tstrachota commented on app/lib/foreman_deployments/validation/remove_ids_visitor.rb:
```diff
@@ -2,7 +2,7 @@ module ForemanDeployments
     class RemoveIdsVisitor
       def visit(subject)
-        if subject.is_a? ForemanDeployments::Tasks::BaseDefinition
+        if subject.is_a? ForemanDeployments::Tasks::CreationTaskDefinition
```
http://projects.theforeman.org/issues/11952

Commit:
Refs #10462 - complexity refactor

ShimShtein commented:
Merged as https://github.com/theforeman/foreman_deployments/commit/bf26477e0b8d668ae89790b7c5f49af54d0a01e1

Thanks @tstrachota
